### ISSUE TITLE
swtpm-localca failing with: Need read/write rights on statedir /var/lib/swtpm-localca for user root.

### ISSUE THREAD

**Issue by klausenbusk** [stefanberger/swtpm#572]
**Describe the bug**
When using `swtpm` with `libvirt` the `swtpm_setup` command fails with:
```
Starting vTPM manufacturing as root:root @ Fri 01 Oct 2021 01:25:09 PM CEST
Successfully created EK.
  Invoking /usr/share/swtpm/swtpm-localca --type ek --ek f80de709b33c0e1592b1e93961cef091a766de7b6039fbc3b785b6621fb4bd34b823e257d634e10f7c71f1174d256a98ee4b8dd8b1cb98b2ee4ac6b412c511fbc3ea3167ed27f31b7bc5f4b28add67143862accfa551eb05708cea11ba1ad77f6e335ce9611f16dfa8de8596ef5d5d5130e243bc4169a8d784fd9d59cfe09c6739475fb43ad39fd2e87b05b13e54a79e5ba83265072e96e5ca0183a020cfafc83c625f95e8214ac5c610565c48fe00c41c4a52655e677cd5d24e60d9a79a8cd54353a1aa215d86f91e3ac45aae5bdd6a6f7262e78341ccb7cfa4b524844001aef2a6351f68815466d97df06f60817b57419ba5baa07c557d9e1ca34a783ffae3 --dir /var/lib/libvirt/swtpm/cbbce4c7-66bc-4547-953b-1e20f393089f/tpm1.2 --logfile /var/log/swtpm/libvirt/qemu/foo-swtpm.log --vmid foo:cbbce4c7-66bc-4547-953b-1e20f393089f --tpm-spec-family 1.2 --tpm-spec-level 2 --tpm-spec-revision 116 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:00740001 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options
Need read/write rights on statedir /var/lib/swtpm-localca for user root.
swtpm-localca exit with status 256:
An error occurred. Authoring the TPM state failed.
Ending vTPM manufacturing @ Fri 01 Oct 2021 01:25:10 PM CEST
```

**To Reproduce**
1. Install `libtpms 0.8.6-1`, `swtpm{,-libs,-tools} 0.6.1-1` from https://launchpad.net/~stefanberger/+archive/ubuntu/swtpm-focal on Debian
2. Invoke `virt-install` with `--tpm emulator`
3. `virt-install` fails with:
    ```
    Starting install...
    ERROR    internal error: Could not run '/usr/bin/swtpm_setup'. exitstatus: 1; Check error log '/var/log/swtpm/libvirt/qemu/foo-swtpm.log' for details.
    Domain installation does not appear to have been successful.
    If it was, you can restart your domain by running:
      virsh --connect qemu:///system start foo
    otherwise, please restart your installation.
    ```

**Expected behavior**
`swtpm-localca` should not fail when running as root or perhaps the permission on `/var/lib/swtpm-localca` should be changed form `750` to `777`?

**Desktop (please complete the following information):**
 - OS: Debian
 - Version: 11 (bullseye)

**Versions of relevant components**
```sh
$ dpkg -l | grep -E 'libtpms|swtpm|openssl|gnutls'
ii  gnutls-bin                           3.7.1-5                        amd64        GNU TLS library - commandline utilities
ii  libcurl3-gnutls:amd64                7.74.0-1.3+b1                  amd64        easy-to-use client-side URL transfer library (GnuTLS flavour)
ii  libgnutls-dane0:amd64                3.7.1-5                        amd64        GNU TLS library - DANE security support
ii  libgnutls30:amd64                    3.7.1-5                        amd64        GNU TLS library - main runtime library
ii  libtpms0:amd64                       0.8.6-1                        amd64        TPM emulation library
ii  openssl                              1.1.1k-1+deb11u1               amd64        Secure Sockets Layer toolkit - cryptographic utility
ii  swtpm                                0.6.1-1                        amd64        Libtpms-based TPM emulator
ii  swtpm-libs:amd64                     0.6.1-1                        amd64        Common libraries for TPM emulators
ii  swtpm-tools                          0.6.1-1                        amd64        Tools for the TPM emulator
```

**Additional context**

```sh
$ ls -ld /var/lib/swtpm-localca/
drwxr-x--- 1 tss root 230  1 okt 13:31 /var/lib/swtpm-localca/
```

I'm unsure if this is a packaging bug, bug in `swtpm-localca` (perhaps it should skip the check when uid == 0?) or a bug in `libvirt`.

**Comment by stefanberger** [comment]
If you are running this as non-root user, please run the script `/usr/share/swtpm/swtpm-create-user-config-files` first to create the config files for your user account.

**Comment by klausenbusk** [comment]
> Who is the owner and what are the permissions on /var/lib/swtpm-localca ?

(`-d list directories themselves, not their contents`)
```sh
$ ls -ld /var/lib/swtpm-localca/
drwxr-x--- 1 tss root 230  1 okt 13:31 /var/lib/swtpm-localca/
```

> Also, what are the contents of /etc/libvirt/qemu.conf related to swtpm:

```
# User for the swtpm TPM Emulator
#
# Default is 'tss'; this is the same user that tcsd (TrouSerS) installs
# and uses; alternative is 'root'
#
#swtpm_user = "tss"
#swtpm_group = "tss"
```

So it should be running as `tss`, hmm..

**Comment by stefanberger** [comment]
I just installed my package on Ubuntu 20.04. Can you try the following as root:

```
mkdir /tmp/myvtpm
swtpm_setup --tpm2 --tpmstate /tmp/myvtpm --create-ek-cert --create-platform-cert
```

It works fine for me...

**Comment by klausenbusk** [comment]
Looks like [it may is configured when `libvirtd` is started](https://github.com/libvirt/libvirt/blob/c8238579fb0b1c3affbd77749ae2b2c4dfafb2d6/src/qemu/qemu_conf.c#L184-L189). I just restarted `libvirtd` and after some cleanup it is now working.

Stupid user error, sorry for the noise :/

**Comment by stefanberger** [comment]
Never mind. I am glad it works for you on Debian.

**Comment by dtbartle** [comment]
Ah, wonderful, worked for me!

**Comment by rjlan** [comment]
I'm getting a similar error on Ubuntu 22.04. I had a working VM named win10 and upgraded to win11. After DARING to rename the VM to win11, I now get an error:

error: internal error: Could not run '/usr/bin/swtpm_setup'. exitstatus: 1; Check error log '/var/log/swtpm/libvirt/qemu/win11-swtpm.log' for details.

Log:

Starting vTPM manufacturing as swtpm:swtpm @ Sat 25 Jun 2022 08:04:00 AM EDT
Successfully created RSA 2048 EK with handle 0x81010001.
  Invoking /usr/lib/x86_64-linux-gnu/swtpm/swtpm-localca --type ek --ek 954b9600d45aba727dc309d63627d5645f96fad24a3fe73673acb867acfcc14095a1adae4b36b74359fcb0bb1afb6db30e14c359a48d7edd7c1ea66f340c5c93840bec306d5e1ad7b82703a80f4d653efb106c2b32b1094cdd7d1ba1b6afed21b1c99a4fc>
Need read/write rights on statedir /var/lib/swtpm-localca for user swtpm.
swtpm-localca exit with status 1:
An error occurred. Authoring the TPM state failed.
Ending vTPM manufacturing @ Sat 25 Jun 2022 08:04:00 AM EDT

> Who is the owner and what are the permissions on /var/lib/swtpm-localca ?

$ ls -l /var/lib/ |grep swtpm-localca
drwxr-x---  2 root          root          4096 Nov  7  2021 swtpm-localca

> Also, what are the contents of /etc/libvirt/qemu.conf related to swtpm:

swtpm_user = "swtpm"
swtpm_group = "swtpm"

**Comment by stefanberger** [comment]
> I'm getting a similar error on Ubuntu 22.04. I had a working VM named win10 and upgraded to win11. After DARING to rename the VM to win11, I now get an error:

So you had the VM working with an attched vTPM ? The renaming of the VM wouldn't have any impact on the read/write rights on that directory. 



>>    Who is the owner and what are the permissions on /var/lib/swtpm-localca ?

>$ ls -l /var/lib/ |grep swtpm-localca
>drwxr-x--- 2 root root 4096 Nov 7 2021 swtpm-localca

>>    Also, what are the contents of /etc/libvirt/qemu.conf related to swtpm:

>swtpm_user = "swtpm"
>swtpm_group = "swtpm"

Did you change any of these since you started your initial Win 10 VM for the first time? What happend on this host that it worked initially but doesn't work any more 'now'? Where did you initially install the swtpm package from?

You may want to try the following command to see whether this resolves the issue:
```
chown -R swtpm:swptm /var/lib/swtpm-localca
```

**Comment by mrmikee** [comment]
## KVM virt-installer  VM create issue due to missing user swtpm

### tl;dr

```bash
# confirm user existed in /etc/passwd
grep swtpm /etc/passwd

# added swtpm user to end of file: /etc/libvirt/qemu.conf
swtpm_user="swtpm"
swtpm_group="swtpm"

# restarted the service
systemctl restart libvirtd
```

------------------------------------------------------------------


**KVM Host:** Ubuntu 22.04 
**KVM Guest:** Ubuntu 20.04 UEFI bios option set 

This happened when creating a new VM with virt-installer with a UEFI bios.  
(In this case a VM import from an Ubuntu Cloud Image)

Host has uefi package `ovmf 2022.02-3ubuntu0.22.04.1`
Host was `apt dist-upgrade` from 18.04 in case that process didn't update the qemu.conf file.

```bash
virt-install \
 ...
 --os-variant=ubuntu20.04 \
 --boot uefi

```
##### Log file error:

```bash
Starting vTPM manufacturing as tss:tss @ Wed 19 Jul 2023 07:17:21 AM EDT
Successfully created RSA 2048 EK with handle 0x81010001.
  Invoking /usr/lib/x86_64-linux-gnu/swtpm/swtpm-localca --type ek --ek b6abcde...8c95a --dir /var/lib/libvirt/swtpm/6a...52a/tpm2 --logfile /var/log/swtpm/libvirt/qemu/myVM-swtpm.log --vmid myVM:6aa...sdfaa --tpm-spec-family 2.0 --tpm-spec-level 0 --tpm-spec-revision 164 --tpm-manufacturer id:00001014 --tpm-model swtpm --tpm-version id:20191023 --tpm2 --configfile /etc/swtpm-localca.conf --optsfile /etc/swtpm-localca.options

Need read/write rights on statedir /var/lib/swtpm-localca for user tss.

swtpm-localca exit with status 1: 
An error occurred. Authoring the TPM state failed.
Ending vTPM manufacturing @ Wed 19 Jul 2023 07:17:21 AM EDT
```
##### ran `stefanberger` suggested fix from [comment 950325420](https://github.com/stefanberger/swtpm/issues/572#issuecomment-950325420)

```bash
/usr/share/swtpm/swtpm-create-user-config-files
```

##### file owner was swtpm:root

```bash
ls -l /var/lib |grep swtpm-localca
drwxr-x---  2 swtpm     root      4096 Jul 19 07:31 swtpm-localca
```

##### QEMU config file had NO swtpm config

```bash
cat /etc/libvirt/qemu.conf |grep swtpm
####nothing returned
```

##### Re-installed packages `swtpm` and `swtpm-tools`

```bash
sudo apt remove --purge swtpm swtpm-tools

sudo apt install swtpm swtpm-tools

#NOTE: this did not fix the issue as I already had the corrected version 0.6.3
sudo dpkg --list |grep swtpm
ii  libtss2-tcti-swtpm0:amd64  3.2.0-1ubuntu1
ii  swtpm                      0.6.3-0ubuntu3.2
ii  swtpm-tools                0.6.3-0ubuntu3.2 
```

##### The actual fix for me

```bash
# confirm user existed in /etc/passwd
grep swtpm /etc/passwd

# added swtpm user to end of file: /etc/libvirt/qemu.conf
swtpm_user="swtpm"
swtpm_group="swtpm"

# restarted the service
systemctl restart libvirtd
```
